File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root:
            return None
        if root.val==p.val or root.val==q.val:
            return root

        leftlca=self.lowestCommonAncestor(root.left,p,q)
        rightlca=self.lowestCommonAncestor(root.right,p,q)

        if leftlca and rightlca:
            return root
        return leftlca or rightlca
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxheight(TreeNode* node){
        if(!node){
            return 0;
        }
        return 1+max(maxheight(node->left),maxheight(node->right));
    }
    int diameterOfBinaryTree(TreeNode* root) {

        if(!root){
            return 0;
        }

        int l=maxheight(root->left);
        int r=maxheight(root->right);
        int dia=l+r;

        int sub=max(diameterOfBinaryTree(root->left),diameterOfBinaryTree(root->right));

        return max(dia,sub);
    
        
    }
};
